a3p-integration/proposals/a:upgrade-next/exitOffer.js

```diff
@@ -0,0 +1,84 @@
+import { parseArgs, promisify } from 'util';
+import { execFile } from 'child_process';
+import { writeFile } from 'fs/promises';
+
+const options = /** @type {const} */ ({
+  id: { type: 'string' },
+  from: { type: 'string' },
+  bin: { type: 'string', default: '/usr/src/agoric-sdk/node_modules/.bin' },
+});
+
+const Usage = `
+Try to exit an offer, reclaiming any associated payments.
+
+  node exitOffer.js --id ID --from FROM [--bin PATH]
+
+Options:
+  --id <offer id>
+  --from <address or key name>
+
+  --bin <path to agoric and agd> default: ${options.bin.default}
+`;
+
+const badUsage = () => {
+  const reason = new Error(Usage);
+  reason.name = 'USAGE';
+  throw reason;
+};
+
+const { stringify: q } = JSON;
+// limited to JSON data: no remotables/promises; no undefined.
+const toCapData = data => ({ body: `#${q(data)}`, slots: [] });
+
+const { entries } = Object;
+/**
+ * @param {Record<string, string>} obj - e.g. { color: 'blue' }
+ * @returns {string[]} - e.g. ['--color', 'blue']
+ */
+const flags = obj =>
+  entries(obj)
+    .map(([k, v]) => [`--${k}`, v])
+    .flat();
+
+const execP = promisify(execFile);
+
+const showAndRun = (file, args) => {
+  console.log('$', file, ...args);
+  return execP(file, args);
+};
+
+const doAction = async (action, from) => {
+  const tmpOffer = await execP('tempfile').then(
+    ({ stdout }) => stdout.split('\n')[0],
+  );
+  await writeFile(tmpOffer, q(toCapData(action)));
+
+  const out = await showAndRun('agoric', [
+    'wallet',
+    ...flags({ 'keyring-backend': 'test' }),
+    'send',
+    ...flags({ offer: tmpOffer, from }),
+  ]);
+  return out.stdout;
```

Comment:
What is "tempfile"? If we're shelling out for this, wouldn't `mktemp` have the broadest support? But better still would be to use [built-in Node.js functionality](https://nodejs.org/docs/latest-v18.x/api/fs.html#fspromisesmkdtempprefix-options)—and to clean up after ourselves when successful.
```suggestion
  const tmpDir = await mkdtemp();
  const tmpOffer = path.join(tmpDir, 'offer.json');
  await writeFile(tmpOffer, q(toCapData(action)));

  const out = await showAndRun('agoric', [
    'wallet',
    ...flags({ 'keyring-backend': 'test' }),
    'send',
    ...flags({ offer: tmpOffer, from }),
  ]);
  await rm(tmpDir, { recursive: true, force: true }).catch(err => console.error(err));
  return out.stdout;
```

Reply:
Thanks.
TIL: node has built-in temp file support.

In agoric-sdk, we depend on various packages for this, so I inferred that node doesn't include it.